### PR TITLE
[ONNX Importer] Add RNN and GRU.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
Add RNN and GRU modules in the ONNX Importer.
This PR is similar to #3713.
I added both RNN and GRU in the same PR because they are very similar.

**Documentation**
None

**Test Plan**
Add ONNX models (and Python generator scripts) with PyTorch numerical references.

### CODE REVIEW

**Comment by mciprian13** [comment]
@jfix71 As promised, I come back with double the fun: RNN and GRU.

File: lib/Graph/Graph.cpp [hunk 1]
```diff
@@ -2912,13 +2912,487 @@ void Function::createLSTM(PlaceholderBindings &bindings,
...
+      bWz = createReshape(prefix + ".bWz.reshape", bWz, {hiddenSize});
+      bWr = createReshape(prefix + ".bWr.reshape", bWr, {hiddenSize});
+      bWh = createReshape(prefix + ".bWh.reshape", bWh, {hiddenSize});
+      bRz = createReshape(prefix + ".bRz.reshape", bRz, {hiddenSize});
+      bRr = createReshape(prefix + ".bRr.reshape", bRr, {hiddenSize});
+      bRh = createReshape(prefix + ".bRh.reshape", bRh, {hiddenSize});
+    }
+
+    // Create H slice for this direction.
+    Node *Hinit = createSlice(prefix + ".H.slice", initial_h.getNode(),
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Graph/Graph.cpp]: Done.

File: lib/Graph/Graph.cpp [hunk 2]
```diff
@@ -2912,13 +2912,487 @@ void Function::createLSTM(PlaceholderBindings &bindings,
...
+
+      bWi = createSlice(prefix + ".bWi.", B, RNN_B_SLICE_RANGE(sliceIdx0, 0));
+      bRi = createSlice(prefix + ".bRi.", B, RNN_B_SLICE_RANGE(sliceIdx0, 1));
+
+      bWi = createReshape(prefix + ".bWi.reshape", bWi, {hiddenSize});
+      bRi = createReshape(prefix + ".bRi.reshape", bRi, {hiddenSize});
+    }
+
+    // Create H slice for this direction.
+    Node *Hinit = createSlice(prefix + ".H.slice", initial_h.getNode(),
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Graph/Graph.cpp]: Done.

File: lib/Graph/Graph.cpp [hunk 3]
```diff
@@ -3124,20 +3595,14 @@ void Function::createONNXLSTM(llvm::StringRef namePrefix, NodeValue X,
     }
 
     // Create H slice for this direction.
-    Node *Hinit = initial_h.getNode();
-    if (numDirections > 1) {
-      Hinit = createSlice(prefix + ".H.slice", Hinit,
-                          LSTM_H_SLICE_RANGE(sliceIdx0));
-    }
+    Node *Hinit = createSlice(prefix + ".H.slice", initial_h.getNode(),
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Graph/Graph.cpp]: Done.

File: lib/Graph/Graph.cpp [hunk 4]
```diff
@@ -3124,20 +3595,14 @@ void Function::createONNXLSTM(llvm::StringRef namePrefix, NodeValue X,
...
     Hinit =
         createReshape(prefix + ".H.reshape", Hinit, {batchSize, hiddenSize});
 
     // Create C slice for this direction.
-    Node *Cinit = initial_c.getNode();
-    if (numDirections > 1) {
-      Cinit = createSlice(prefix + ".C.slice", Cinit,
-                          LSTM_C_SLICE_RANGE(sliceIdx0));
-    }
+    Node *Cinit = createSlice(prefix + ".C.slice", initial_c.getNode(),
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Graph/Graph.cpp]: Done.

File: tests/unittests/OnnxExporterTest.cpp [hunk 5]
```diff
@@ -124,6 +124,16 @@ TEST(exporter, onnxModels) {
       llvm::outs() << "Ignore output file: " << name << "\n";
       continue;
     }
+    if (name.find("rnn") != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OnnxExporterTest.cpp]: The models don not work when reloading (they are loaded and written without errors but when reloading some errors pop up). I debugged a little bit but I can`t figure out what is the problem. The errors look like this:
```
Error message: could not find constant with name lstm_Ro_transp
Error message: No node under name lstm_Y_c
```
Can anyone help with this? Or maybe raise a separate issue?

**Comment by mciprian13** [comment]
- I updated the little nits you mentioned.
- I refactored a little bit the code from OnnnModelLoader.cpp to reuse more code between the 3 modules (the part when reading the direction and the one reading the activations from the proto).
- As a bonus I also added the "input_forget" feature for the LSTM (plus unit test).
- The rest of the code I don`t think we can safely refactor for code reuse since the 3 modules (RNN, GRU, LSTM) might evolve slightly different in the future.
- The remaining problem is the one related to the ONNX Exporter test which I cannot figure out.

File: tests/unittests/OnnxExporterTest.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfix71) [reply tests/unittests/OnnxExporterTest.cpp]: Ok, yeah we need to improve the testing for the exporter, this can be done separately, thanks.

**Pull request closed without merging**